core/domain/learner_group_services.py

```diff
@@ -0,0 +1,565 @@
+    learner_group_model.put()
+
+    learner_group = get_learner_group_from_model(learner_group_model)
+    learner_group.validate()
+
+    if len(learner_group_model.invited_student_user_ids) > 0:
+        invite_students_to_learner_group(
+            group_id, learner_group_model.invited_student_user_ids)
+
+    return learner_group
+
+
+def update_learner_group(
+    group_id: str,
+    title: str,
+    description: str,
+    facilitator_user_ids: List[str],
+    student_ids: List[str],
+    invited_student_ids: List[str],
+    subtopic_page_ids: List[str],
+    story_ids: List[str]
+) -> learner_group_domain.LearnerGroup:
+    """Updates a learner group if it is present.
+
+    Args:
+        group_id: str. The id of the learner group to be updated.
+        title: str. The title of the learner group.
+        description: str. The description of the learner group.
+        facilitator_user_ids: str. List of user ids of the facilitators of the
+            learner group.
+        student_ids: list(str). List of user ids of the students of the
+            learner group.
+        invited_student_ids: list(str). List of user ids of the students who
+            have been invited to join the learner group.
+        subtopic_page_ids: list(str). The ids of the subtopics pages that are
+            part of the learner group syllabus. Each subtopic page id is
+            represented as a topicId:subtopicId string.
+        story_ids: list(str). The ids of the stories that are part of the
+            learner group syllabus.
+
+    Returns:
+        learner_group: learner_group_domain.LearnerGroup. The domain object
+        of the updated learner group.
+    """
+
+    learner_group_model = learner_group_models.LearnerGroupModel.get(
+        group_id, strict=True
+    )
+
+    old_invited_student_ids = set(learner_group_model.invited_student_user_ids)
+    new_invited_student_ids = set(invited_student_ids)
+    if new_invited_student_ids != old_invited_student_ids:
+        newly_added_invites = list(
+            new_invited_student_ids - old_invited_student_ids
+        )
+        newly_removed_invites = list(
+            old_invited_student_ids - new_invited_student_ids
+        )
+        invite_students_to_learner_group(
+            group_id, newly_added_invites)
+        remove_invited_students_from_learner_group(
+            group_id, newly_removed_invites)
+
+    learner_group_model.title = title
+    learner_group_model.description = description
+    learner_group_model.facilitator_user_ids = facilitator_user_ids
+    learner_group_model.student_user_ids = student_ids
+    learner_group_model.invited_student_user_ids = invited_student_ids
+    learner_group_model.subtopic_page_ids = subtopic_page_ids
+    learner_group_model.story_ids = story_ids
+
+    learner_group_model.update_timestamps()
+    learner_group_model.put()
+
+    learner_group = get_learner_group_from_model(learner_group_model)
+    learner_group.validate()
```

Comment:
Do the validate before the put, since we don't want invalid models in database.

Reply:
done